sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -339,6 +341,28 @@ object EliminateDistinct extends Rule[LogicalPlan] {
   }
 }
 
+/**
+ * Remove useless FILTER clause for aggregate expressions.
+ * This rule should be applied before RewriteDistinctAggregates.
+ */
+object EliminateAggregateFilter extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan = plan transformExpressions  {
+    case ae @ AggregateExpression(_, _, _, Some(Literal.TrueLiteral), _) =>
+      ae.copy(filter = None)
+    case AggregateExpression(af: DeclarativeAggregate, _, _, Some(Literal.FalseLiteral), _) =>
+      val initializer = SafeProjection.create(af.initialValues)
+      val evaluator = SafeProjection.create(af.evaluateExpression :: Nil, af.aggBufferAttributes)
+      val buffer = new SpecificInternalRow(af.aggBufferAttributes.map(_.dataType))
+      val initBuffer = initializer(buffer).copy()
```

Comment:
why do we need to copy?

Reply:
Looking at `AggregationIterator.initializeBuffer`, I think we should do
```
val initProj = MutableProjection.create...
val buffer = new SpecificInternalRow(af.aggBufferAttributes.map(_.dataType))
initProj.target(buffer)(EmptyRow)
```

Reply:
no need copy.